### DISCUSSION TITLE
Akka Startup Spec

### DISCUSSION THREAD

**Discussion by Zetanova** [akkadotnet/akka.net#5447] [Ideas]
The current akka startup (including remote and cluster) is depends currently heavily on thread-locks to synchronize. 
This happens most of the time over `Lazy<Extension>.Value` of `AkkaSystem._extensions`.

Because the caller of a lazy loaded extension is most of the time an actor, the call sequence is none deterministic. 
An other issue is that a hit on the lock of `Lazy<Extension>.Value` produces a relative long lock period. 

We would need to decide what startup features are required and then how we should implement them.

Here some ideas from my side:
- Support of AkkaSystem.StartAsync and AkkaSystem.StopAsync 
- Inducing of an Interface to mark a component async initializable
- A mechanisms to check if an extension was initialized

One idea would be to create an `ExtensionManager` actor, 
that would lazy create/supervise all extensions and publish an `ExtensionLoaded` event on the EventStream.
All other components could subscribe and/or request extensions similar to the current ClusterState Events.

**Comment by Aaronontheweb** [comment]
I need to respond to this - sorry for the delay. Had _a lot_ going on over the holidays and during the start of 2022 thus far.

**Comment by eaba** [comment]
Can I call this "**Akka Improvement Proposal (AIP)**"?

@Zetanova I like this idea. Please can you come up with more detail as follows:

- Approach: how do we approach the implementation.
- API Changes: API changes we need to be aware of with code samples
- Internal Changes: What are the internal changes with code samples
- Compatibility: Will this introduce any compatibility issue
- Tests plan: What are the test plan to verify changes with code sample. Unit tests and integration tests!

Any idea that will help to improve **Akka.NET** is always welcomed, however, if you provide more detail, it will help the team in accepting and working on it.

Please include additional code samples where possible.

**Reply by Zetanova** [reply]
This Points are exactly what I want to discuss here. 
and to create a AIP from the Ideas collected.

**Comment by eaba** [comment]
Do you have any idea on how we can approach this? Let start with that first. @Aaronontheweb
